R/sampling.R

```diff
@@ -159,3 +159,161 @@ sample_combinations <- function(ntrain, ntest, nsamples, joint_sampling = TRUE)
 
   return(ret)
 }
+
+#' Sample ctree variables from a given conditional inference tree
+#'
+#' @param tree ctree built from the party package. This includes the indices of the features to condition upon.
+#' @param n_samples Numeric. Indicates how many samples to use for MCMC.
+#' @param x_test Matrix, data.frame or data.table with the features of the observation whose
+#' predictions ought to be explained (test data). Dimension \code{1xp} or \code{px1}.
+#' @param x_train Matrix, data.frame or data.table with training data.
+#' @param p Positive integer. The number of features.
+#'
+#' @inheritParams global_arguments
+#'
+#' @return data.table with \code{n_samples} (conditional) Gaussian samples
+#'
+#' @keywords internal
+#'
+#' @examples
+#' # TODO: Add simple example
+#'
+#' @author Annabelle Redelmeier
+
+sample_ctree <- function(tree,
+                         n_samples,
+                         x_test,
+                         x_train,
+                         p) {
+
+
+  datact <- tree$tree
+  # print(tree$given_ind)
+
+  cnms <- colnames(x_test)
+  if (length(tree$given_ind) %in% c(0, p)) {
+    ret <- matrix(x_test, ncol = p, nrow = 1)
+  } else {
+    given_ind <- tree$given_ind
+    given_ind.vec <- rep(0, length(x_test))
+    given_ind.vec[given_ind] <- 1
+
+    dependent_ind <- tree$dependent_ind
+
+    x_test_given <- x_test[given_ind]
+
+    xp <- data.table(matrix(x_test_given, nrow = 1, ncol = length(x_test_given)))
+    colnames(xp) <- paste0("V", given_ind) # this is important for where() below
+
+    # x0 <- x_test
+    # dependent_ind <- (1:dim(x_train)[2])[-given.ind.vec]
+    # X_given <- x_test[given.ind]
+
+    # xp <- data.table(matrix(x_test, nrow = 1, ncol = length(X_given)))
+    # colnames(xp) <- paste0("V", given.ind)
+
+    fit.nodes <- where(datact)
+    nodes <- unique(fit.nodes)
+    no.nodes <- length(nodes)
+    pred.nodes <- where(object = datact, newdata = xp) ## newdata must be a data.frame and have the same colnames as x
+
+    rowno <- 1:dim(x_train)[1]
+    # newrowno <- vector("integer", n_samples)
+
+    newrowno <- sample(rowno[fit.nodes == pred.nodes], n_samples, replace = TRUE)
+
+    depDT <- data.table(matrix(x_train[newrowno, dependent_ind], ncol = length(dependent_ind)))
+    givenDT <- data.table(matrix(x_test[1, given_ind], ncol = length(given_ind)))
+
+    ret <- data.table(matrix(0, nrow = n_samples, ncol = length(x_test)))
+    ret[, paste0("V", dependent_ind) := depDT]
+    ret[, paste0("V", given_ind) := givenDT]
```

Comment:
I think this could be done in a similar way as impute_observations_cpp is used in impute_observations, see

https://github.com/NorskRegnesentral/shapr/blob/78dde469adb1bb737717b02a9d1bc8326a6784ec/R/observations.R#L22
https://github.com/NorskRegnesentral/shapr/blob/78dde469adb1bb737717b02a9d1bc8326a6784ec/src/impute_data.cpp#L39